### PR TITLE
KAFKA-7192: Wipe out state store if EOS is turned on and checkpoint file does not exist

### PR DESCRIPTION
**Pull request by mjsax**
Unified backport of PRs #5421 and #5430

Also compare backport to `0.11.0` #5641 and `1.0` #5657

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @bbejeck @vvcephei

**Review by guozhangwang — COMMENTED** [review]
I made a pass on https://github.com/apache/kafka/pull/5768/files, assuming this PR did not introduce additional logic to it, lgtm once 5768 is merged.

**Comment by mjsax** [comment]
Updates this to include the fix from #5915

Triggered system test: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2078/

**Comment by mjsax** [comment]
Retriggered failing system test: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2080/ (passed)

Retriggered again: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2082/

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:70]: I have some questions just for my education.

I think I get it what the issue is.  The `ProcessorStateManager#register` creates a new `StateRestorer` each time, and it gets passed in here, so without reusing the `existingRestorer` restoring never completes. Is that correct?  My question is `StateRestorer` should contain the same state even if a new instance, so how does using the existing restorer solve the issue? From what I can see the state should be the same between the new and existing restorer.  Since the fix seems to be the same as #5915, I'll ask my question here.

**Review by bbejeck — APPROVED** [review]
Thanks for the fix @mjsax I have one question for my own education, otherwise LGTM

**Comment by mjsax** [comment]
System test passed. Re-triggered as: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2083/

**Comment by mjsax** [comment]
Older branch not compatible with Java10 or Java7...

**Review by guozhangwang — APPROVED** [review]

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:70]: >  Is that correct?

Yes.

For EOS, on startup, we read the checkpoint file, store the information in the restorer, and delete the checkpoint file (we only write one on clean shutdown). The magic happens in L173/180 `if (restorer.checkpoint() != StateRestorer.NO_CHECKPOINT) {` and L204 `restorer.setCheckpointOffset(restoreConsumer.position(restoringPartition));`. 

If there is no checkpoint store in the restorer, we seekToBeginning and mark as `needsPositionUpdate`. Later, we wipe out the store and put the current beginningOffset as checkpoint to the restorer. Thus, when we re-register, we need to preserve this checkpoint information. The newly created `StateRestored` would get initialized by looking if there is checkpoint file on disk and would be initialized with "no checkpoint".

It's a little hacky... Maybe we can revisit this design at some point and make it cleaner.

Makes sense?

**Comment by mjsax** [comment]
System test passed. Triggered one more run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2088/

(investigating first failing test tun)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:70]: Yes, thanks for the clarification.

**Pull request merged**
